README.md
<h1 align=center>
Script Template
</h1>

<p align=center>  
  <a href="LICENSE">
    <img src="https://img.shields.io/github/license/mygamedevtools/script-template" />
  </a>
  <a href="https://github.com/mygamedevtools/script-template/actions/workflows/release.yml">
    <img src="https://github.com/mygamedevtools/script-template/actions/workflows/release.yml/badge.svg" />
  </a>
  <a href="https://github.com/mygamedevtools/script-template/commits/">
    <img src="https://img.shields.io/github/last-commit/mygamedevtools/script-template" />
  </a>
</p>

<p align=center>
  <a href="https://openupm.com/packages/com.mygamedevtools.script-template/">
    <img src="https://img.shields.io/npm/v/com.mygamedevtools.script-template?label=openupm&registry_uri=https://package.openupm.com" />
  </a>
  <a href="https://github.com/mygamedevtools/script-template/releases/latest">
    <img src="https://img.shields.io/github/v/release/mygamedevtools/script-template?sort=semver" />
  </a>
  <a href="https://github.com/semantic-release/semantic-release">
    <img src="https://img.shields.io/badge/semantic--release-angular-e10079?logo=semantic-release" />
  </a>
</p>

<p align=center><i>
A simple tool to add and manage more script templates.
</i></p>

What's this about?
---

This tool adds custom Script Templates to Unity that will serve as alternatives for the default \"C#Script\" that I personally always modify before writing my scripts. In essence, it allows customization of the script templates with dynamic data such as signature and namespace settings. Additionally, it adds templates for others types of scripts rather than just Mono Behaviours. You can customize the templates after they have been copied as well.

Installation
---

### OpenUPM

This package is available on the [OpenUPM](https://openupm.com/packages/com.mygamedevtools.script-template) registry. Add the package via the [openupm-cli](https://github.com/openupm/openupm-cli):

```
openupm add com.mygamedevtools.script-template
```

### [Installing from Git](https://docs.unity3d.com/Manual/upm-ui-giturl.html) _(requires [Git](https://git-scm.com/) installed and added to the PATH)_

1. Open `Edit/Project Settings/Package Manager`.
2. Click <kbd>+</kbd>.
3. Select `Add package from git URL...`.
4. Paste `com.mygamedevtools.script-template` into name.
5. Click `Add`.

Usage
---

![First Image](https://user-images.githubusercontent.com/9505905/111533338-6dd37580-8745-11eb-968d-37102b4b5e5c.png)

From `Assets/Script Templates Editor` you can open the **Script Templates Editor** Window.

![Custom Script Template Editor](https://github.com/mygamedevtools/script-template/assets/9505905/04d70654-95b9-4ab2-83f2-33e23a925631)

Here you can preview in realtime what your script will look like, customize the values and copy the templates. 

### Settings

You can decide to enable both the Signature and the Namespace modules, each with their own set of custom settings.

#### Signature

The Signature module adds a simple signature header on your script following the template:

```cs
/**
 * #SCRIPTNAME#.cs
 * Created by: #AUTHOR#
 * Created on: #CREATIONDATE#
 */
```

The `#SCRIPTNAME#` is no stranger to custom script templates, as it is replaced by the default Unity asset processor.
However, both the `#AUTHOR#` and `#CREATIONDATE#` will be replaced by the values you set on the `Author`, `Email(optional)` and `Use Local Date` fields under the Namespace module group.

#### Namespace

The Namespace module adds a namespace to your file, indenting the template content under it.
For that to work, you need to have a template such as:

```
#NAMESPACE#public class #SCRIPTNAME# : MonoBehaviour
{
# #public void Start()
# #{
# ## #
# #}
}
```

Then, if enabled, the namespace will be replaced according to the values you set.
If you enable `Use Assembly Definition`, make sure the script is in the correct location to be linked to an assembly definition, or the `Default Namespace` will be applied.
Otherwise, if you decide to just use the `Default Namespace`, then it will be applied to all scripts.

#### Indentation

The indentation whitespaces must be marked with the `# #` keyword, one for each indent level, so they can be converted to tabs or spaces and by how much. So, using the previous example:

```
#NAMESPACE#public class #SCRIPTNAME# : MonoBehaviour
{
# #public void Start()
# #{
# ## #
# #}
}
```

By selecting **4 spaces** as the indentation settings, the final created script would be:

```cs
namespace MyNamespace
{
    public class MyClass : MonoBehaviour
    {
        public void Start()
        {

        }
    }
}
```

Finally, you should `Save` your changes once you're satisfied with your settings, to make sure they will be used whenever creating a new script.
You also have the option to `Clear` those settings at any time, if you wish.
This data is saved to the [Editor Prefs](https://docs.unity3d.com/ScriptReference/EditorPrefs.html) cache.


### Templates

This section shows a list of all included Script Templates and you can select them to locate the file.
You also have two buttons: `Copy to Project` and `Copy to Editor`.
The first one will copy all script templates to the current project `Assets/ScriptTemplates` path, limiting the script template usage to that project.
The other will copy all script templates to your Editor folder, and you will be able to use the templates
in all projects.
However if you don't have this package in a project, it will not replace the custom keywords on the file and might lead to compilation issues.
Also, if you update the Unity Editor version, you will lose the script templates and require to copy them again. This is not necessary if you copy them to your project.
Both actions of copying to project and to editor will require an editor restart to take effect.

#### Included Templates:

* Game Objects
  * [Mono Behaviour](https://docs.unity3d.com/ScriptReference/MonoBehaviour.html)
  * [Scriptable Object](https://docs.unity3d.com/ScriptReference/ScriptableObject.html)
  * [State Machine Behaviour](https://docs.unity3d.com/ScriptReference/StateMachineBehaviour.html)
  * [Property Attribute](https://docs.unity3d.com/ScriptReference/PropertyAttribute.html)
* Editor
  * [Editor Window](https://docs.unity3d.com/ScriptReference/EditorWindow.html)
  * [Custom Editor Inspector](https://docs.unity3d.com/ScriptReference/Editor.html)
  * [Property Drawer](https://docs.unity3d.com/ScriptReference/PropertyDrawer.html)
* UI Toolkit
  * [Custom Element](https://docs.unity3d.com/Manual/UIE-expose-custom-control-to-uxml.html)
* Entities
  * [System](https://docs.unity3d.com/Packages/com.unity.entities@1.0/manual/systems-isystem.html)
  * [Component](https://docs.unity3d.com/Packages/com.unity.entities@1.0/manual/components-unmanaged.html)
  * [Authoring](https://docs.unity3d.com/Packages/com.unity.entities@1.0/manual/baking-baker-overview.html) (Baker Workflow)
  * [Managed Component](https://docs.unity3d.com/Packages/com.unity.entities@1.0/manual/components-managed.html)
  * [Managed System](https://docs.unity3d.com/Packages/com.unity.entities@1.0/manual/systems-systembase.html)
* Plain C#
  * Class
  * Struct
  * Enum
  * Interface

How it works
---

The custom Script Templates have custom keywords that will be replaced by the data you entered on the Editor Window and the present date.
The keyword recognition and replacing is done by the [ScriptKeywordReplacer](Assets/CustomScriptTemplate/Editor/ScriptKeywordReplacer.cs) class.

Customization
---

You can customize your Script Templates to suit your needs, for example adding default methods or even new keywords to process.
Just make sure you add them to the correct folder and add the custom keywords following the same rules as the built-in templates.
Keep in mind that if you ever want a feature that it does not have, I may want to include it on this package.

But, why?
---

It all started when I wanted to have my scripts signed so everyone knows who to ask when they have doubts.
Unfortunately with every new version of Unity, I had to redo the tedious process of updating my Script Templates manually.
I'd very much like to be able to [manage my Script Templates over the Unity Hub](https://forum.unity.com/threads/feature-request-manage-script-templates.532962/), but it's a very minor feature.

---

Don't hesitate to create [issues](https://github.com/mygamedevtools/script-template/issues) for suggestions and bugs. Have fun!


CHANGELOG.md
# [3.5.0](https://github.com/mygamedevtools/script-template/compare/3.4.1...3.5.0) (2023-08-24)


### Features

* add more script templates ([#13](https://github.com/mygamedevtools/script-template/issues/13)) ([fb16353](https://github.com/mygamedevtools/script-template/commit/fb16353643b2f0c4f9c56300f4ff719ce222df17))

## [3.4.1](https://github.com/mygamedevtools/script-template/compare/3.4.0...3.4.1) (2023-08-19)


### Bug Fixes

* update indentation replacements with tabs or spaces ([#12](https://github.com/mygamedevtools/script-template/issues/12)) ([05f5ce8](https://github.com/mygamedevtools/script-template/commit/05f5ce891ef146ab3f3baf1c580ad0a13c3e6f96))

# [3.4.0](https://github.com/mygamedevtools/script-template/compare/3.3.0...3.4.0) (2023-03-25)


### Features

* add entities script templates ([2f34568](https://github.com/mygamedevtools/script-template/commit/2f34568c77c0afb266129b98251f75256b81e5d4))

# [3.3.0](https://github.com/mygamedevtools/script-template/compare/3.2.1...3.3.0) (2022-10-25)


### Features

* Add CI configuration ([079c51e](https://github.com/mygamedevtools/script-template/commit/079c51e92b5790ac79f2687381bd0daffed02582))

## [3.2.1] - 2022-10-24
- Added: OpenUPM documentation.

## [3.2.0] - 2022-10-03
- Fixed: Updated assembly definition and namespaces names to reflect the organization name changes.

## [3.1.0] - 2022-10-03
- Changed: Updated organization name to comply with [Unity Package Guidelines](https://unity.com/legal/terms-of-service/software/package-guidelines).

## [3.0.0] - 2022-10-03
- Added: Namespace settings module.
- Added: Live preview of namespace and signature changes.
- Changed: Cached data is now a single entry instead of multiple entries in EditorPrefs.

## [2.2.0] - 2022-09-13
- Added: Support for ISO date format by default.

## [2.1.0] - 2022-09-13
- Changed: Moved repository to My Unity Tools organization
- Changed: Updated package name and author.

## [2.0.1] - 2021-11-29
- Changed: Added support for Unity 2021.2.

## [2.0.0] - 2021-03-17
- Changed: Updated UI screens to use the UI Toolkit.
- Added: Support for multiple script templates.

## [1.3.0] - 2021-02-22
- Changed: Refactored scripts to the My Unity Tools namespace.

## [1.2.2] - 2020-09-11
- Fixed: Script templates not being copied to editor folder.

## [1.2.1] - 2020-09-11
- Added: Reintroduced .meta files.

## [1.2.0] - 2020-09-08
- Removed: .meta files.
- Removed: Asset Store support.

## [1.1.1] - 2020-03-22
- Fixed: script template editing.

## [1.1.0] - 2019-09-26
- Changed: Updated structure to adapt to the Unity Package Manager.

## [1.0.1] - 2019-08-18
- Added: Assembly definition to compile the scripts.

## [1.0.0] - 2019-07-27
- Initial Asset Store release.

[3.2.1]: https://github.com/mygamedevtools/script-template/compare/3.2.0...3.2.1
[3.2.0]: https://github.com/mygamedevtools/script-template/compare/3.1.0...3.2.0
[3.1.0]: https://github.com/mygamedevtools/script-template/compare/3.0.0...3.1.0
[3.0.0]: https://github.com/mygamedevtools/script-template/compare/2.2.0...3.0.0
[2.2.0]: https://github.com/mygamedevtools/script-template/compare/2.1.0...2.2.0
[2.1.0]: https://github.com/mygamedevtools/script-template/compare/2.0.1...2.1.0
[2.0.1]: https://github.com/mygamedevtools/script-template/compare/2.0.0...2.0.1
[2.0.0]: https://github.com/mygamedevtools/script-template/compare/1.3.0...2.0.0
[1.3.0]: https://github.com/mygamedevtools/script-template/compare/1.2.2...1.3.0
[1.2.2]: https://github.com/mygamedevtools/script-template/compare/1.2.1...1.2.2
[1.2.1]: https://github.com/mygamedevtools/script-template/compare/1.2.0...1.2.1
[1.2.0]: https://github.com/mygamedevtools/script-template/compare/1.1.1...1.2.0
[1.1.1]: https://github.com/mygamedevtools/script-template/compare/1.1.0...1.1.1
[1.1.0]: https://github.com/mygamedevtools/script-template/compare/1.0.1...1.1.0
[1.0.1]: https://github.com/mygamedevtools/script-template/compare/1.0.0...1.0.1
[1.0.0]: https://github.com/mygamedevtools/script-template/compare/087dd6c...1.0.0


Editor/ScriptTemplatesEditor.cs
/**
 * ScriptTemplatesEditor.cs
 * Created by: João Borks [joao.borks@gmail.com]
 * Created on: 2019-07-13
 */

using MyGameDevTools.ScriptTemplates.UIToolkit;
using System.IO;
using System.Linq;
using UnityEditor;
using UnityEditor.SceneManagement;
using UnityEditor.UIElements;
using UnityEngine;
using UnityEngine.UIElements;

namespace MyGameDevTools.ScriptTemplates
{
    public class ScriptTemplatesEditor : EditorWindow
    {
        public const string PackageRootPath = "Packages/com.mygamedevtools.script-template/";
        public const string LocalTemplatesPath = "Assets/ScriptTemplates/";

        /// <summary>
        /// Editor path for storing script templates
        /// </summary>
        public static readonly string EditorTemplatesPath = EditorApplication.applicationPath.Replace("Unity.exe", "") + "/Data/Resources/ScriptTemplates/";

        const string _editorRestartMesssage = "Copying Script Templates will requre Editor restart in order to apply the changes. Do you want to restart now?";
        const string _previewTemplate =
            "<color=#608B4E>#SIGNATURE#</color><color=#569cd6>using</color> UnityEngine;\n" +
            "\n" +
            "#NAMESPACE#<color=#569cd6>public class</color> <color=#4ec9b0>#SCRIPTNAME#</color> : <color=#4ec9b0>MonoBehaviour</color>\n" +
            "{\n" +
            "    <color=#569cd6>public void</color> <color=#dcdcaa>Start</color>()\n" +
            "    {\n" +
            "    }\n" +
            "}";

        ScriptTemplateSettings _cachedTemplateSettings;
        ScriptTemplateSettings _previewTemplateSettings;
        ScriptKeywordReplacer _keywordReplacer;

        [MenuItem("Assets/Script Templates Editor", false, 800)]
        public static void ShowWindow() => GetWindow<ScriptTemplatesEditor>("Script Templates").minSize = new Vector2(400, 650);

        void OnEnable()
        {
            var visualTree = AssetDatabase.LoadAssetAtPath<VisualTreeAsset>(PackageRootPath + "UIToolkit/ScriptTemplateWindow.uxml");
            rootVisualElement.Add(visualTree.CloneTree());

            _cachedTemplateSettings = ScriptTemplateSettings.FromEditorPrefs();
            _previewTemplateSettings = new ScriptTemplateSettings()
            {
                Signature = _cachedTemplateSettings.Signature,
                Namespace = _cachedTemplateSettings.Namespace
            };
            _keywordReplacer = new ScriptKeywordReplacer(_previewTemplateSettings);

            BindSettingsView();
            BindTemplateView();
        }

        void BindSettingsView()
        {
            var previewLabel = rootVisualElement.Q<Label>("preview-template");
            var saveButton = rootVisualElement.Q<Button>("save-button");
            var clearButton = rootVisualElement.Q<Button>("clear-button");

            var signatureModule = rootVisualElement.Q<Module>("signature-module");
            signatureModule.SetValueWithoutNotify(_previewTemplateSettings.Signature.Enabled);

            var authorField = signatureModule.Q<TextField>("author");
            authorField.SetValueWithoutNotify(_previewTemplateSettings.Signature.AuthorName);

            var emailField = signatureModule.Q<TextField>("email");
            emailField.SetValueWithoutNotify(_previewTemplateSettings.Signature.AuthorEmail);

            var localDateToggle = signatureModule.Q<Toggle>("localdate");
            localDateToggle.SetValueWithoutNotify(_previewTemplateSettings.Signature.UseLocalDateFormat);

            var namespaceModule = rootVisualElement.Q<Module>("namespace-module");
            namespaceModule.SetValueWithoutNotify(_previewTemplateSettings.Namespace.Enabled);

            var assembly = namespaceModule.Q<Toggle>("assembly");
            assembly.SetValueWithoutNotify(_previewTemplateSettings.Namespace.UseAssemblyDefinition);

            var defaultNamespace = namespaceModule.Q<TextField>("default");
            defaultNamespace.SetValueWithoutNotify(_previewTemplateSettings.Namespace.DefaultNamespace);

            var indentType = namespaceModule.Q<EnumField>("indent-type");
            indentType.SetValueWithoutNotify(_previewTemplateSettings.Namespace.IndentPattern);

            var indentMult = namespaceModule.Q<SliderInt>("indent-mult");
            indentMult.SetValueWithoutNotify(_previewTemplateSettings.Namespace.IndentMultiplier);

            clearButton.SetEnabled(!_cachedTemplateSettings.IsEmpty());
            clearButton.clicked += () =>
            {
                _cachedTemplateSettings.Signature = new SignatureSettings();
                _cachedTemplateSettings.DeleteEditorPrefs();
                saveButton.SetEnabled(_cachedTemplateSettings != _previewTemplateSettings);
                clearButton.SetEnabled(false);
            };

            signatureModule.RegisterValueChangedCallback(e => updatePreview());
            authorField.RegisterValueChangedCallback(e => updatePreview());
            emailField.RegisterValueChangedCallback(e => updatePreview());
            localDateToggle.RegisterValueChangedCallback(e => updatePreview());

            namespaceModule.RegisterValueChangedCallback(e => updatePreview());
            assembly.RegisterValueChangedCallback(e => updatePreview());
            defaultNamespace.RegisterValueChangedCallback(e => updatePreview());
            indentType.RegisterValueChangedCallback(e => updatePreview());
            indentMult.RegisterValueChangedCallback(e => updatePreview());

            saveButton.SetEnabled(false);
            saveButton.clicked += () =>
            {
                _cachedTemplateSettings.Signature = _previewTemplateSettings.Signature;
                _cachedTemplateSettings.Namespace = _previewTemplateSettings.Namespace;
                _previewTemplateSettings.SaveToEditorPrefs();
                saveButton.SetEnabled(false);
                clearButton.SetEnabled(!_cachedTemplateSettings.IsEmpty());
            };

            updatePreview();

            void updatePreview()
            {
                _previewTemplateSettings.Signature = new SignatureSettings
                {
                    Enabled = signatureModule.value,
                    AuthorName = authorField.value,
                    AuthorEmail = emailField.value,
                    UseLocalDateFormat = localDateToggle.value
                };
                _previewTemplateSettings.Namespace = new NamespaceSettings
                {
                    Enabled = namespaceModule.value,
                    UseAssemblyDefinition = assembly.value,
                    DefaultNamespace = defaultNamespace.value,
                    IndentPattern = (IndentPattern)indentType.value,
                    IndentMultiplier = indentMult.value
                };
                previewLabel.text = _keywordReplacer.ProcessScriptTemplate(_previewTemplate, "Assets/Scripts/ExampleScript.cs").Replace("namespace", "<color=#569cd6>namespace</color>").Replace(" ", "<color=#144852>·</color>").Replace("\t", "<color=#144852>→   </color>");
                saveButton.SetEnabled(_cachedTemplateSettings != _previewTemplateSettings && _previewTemplateSettings.IsValid());
            }
        }

        void BindTemplateView()
        {
            var templates = GetPackageScriptTemplates();
            var listView = rootVisualElement.Q<ListView>("template-list");
            listView.makeItem = () =>
            {
                var objectField = new ObjectField(string.Empty)
                {
                    allowSceneObjects = false,
                    objectType = typeof(TextAsset)
                };
                objectField.SetEnabled(false);
                return objectField;
            };
            listView.bindItem = (element, i) => (element as ObjectField).SetValueWithoutNotify(templates[i]);
            listView.itemsSource = templates;
#if UNITY_2021_2_OR_NEWER
            listView.fixedItemHeight = 20;
#else
            listView.itemHeight = 20;
#endif
            listView.selectionType = SelectionType.Single;
#if UNITY_2022_2_OR_NEWER
            listView.selectionChanged += selection => Selection.activeObject = (UnityEngine.Object)selection.FirstOrDefault();
#elif UNITY_2021_2_OR_NEWER
            listView.onSelectionChange += selection => Selection.activeObject = (UnityEngine.Object)selection.FirstOrDefault();
#else
            listView.onSelectionChanged += selection => Selection.activeObject = (UnityEngine.Object)selection.FirstOrDefault();
#endif

            var projectButton = rootVisualElement.Q<Button>("copy-project-button");
            projectButton.clicked += () =>
            {
                var shouldRestart = EditorUtility.DisplayDialog("Editor Restart", _editorRestartMesssage, "Yes", "No");
                var length = templates.Length;
                if (!AssetDatabase.IsValidFolder(LocalTemplatesPath))
                    AssetDatabase.CreateFolder("Assets", "ScriptTemplates");
                for (int i = 0; i < length; i++)
                    AssetDatabase.CopyAsset(AssetDatabase.GetAssetPath(templates[i]), LocalTemplatesPath + templates[i].name + ".txt");
                if (shouldRestart && EditorSceneManager.SaveCurrentModifiedScenesIfUserWantsTo())
                    EditorApplication.OpenProject(Path.Combine(Application.dataPath, ".."));
            };

            var editorButton = rootVisualElement.Q<Button>("copy-editor-button");
            editorButton.clicked += () =>
            {
                var shouldRestart = EditorUtility.DisplayDialog("Editor Restart", _editorRestartMesssage, "Yes", "No");
                var length = templates.Length;
                for (int i = 0; i < length; i++)
                    File.Copy(AssetDatabase.GetAssetPath(templates[i]), EditorTemplatesPath + templates[i].name + ".txt");
                if (shouldRestart && EditorSceneManager.SaveCurrentModifiedScenesIfUserWantsTo())
                    EditorApplication.OpenProject(Path.Combine(Application.dataPath, ".."));
            };
        }

        /// <summary>
        /// Gets all Script Templates included in this package
        /// </summary>
        TextAsset[] GetPackageScriptTemplates()
        {
            var files = Directory.GetFiles(PackageRootPath + "ScriptTemplates/", "*.txt");
            var assets = new TextAsset[files.Length];
            var length = files.Length;
            for (int i = 0; i < length; i++)
                assets[i] = AssetDatabase.LoadAssetAtPath<TextAsset>(files[i]);
            return assets;
        }
    }
}